Guide for observing a contract_call predicate with Chainhook and extracting data from its payload (#723)
hirosystems/docs: content/docs/stacks/chainhook/guides/observing-contract-calls.mdx

---
title: Observing Contract Calls with Chainhook

> description: Learn how to use Chainhook to observe contract calls on the Stacks blockchain

Reviewer: ```suggestion
description: Learn how to use Chainhook to observe smart contract calls.
```

---

import { Step, Steps } from "fumadocs-ui/components/steps";

import SimpleVote from "../resources/simple-vote.mdx";

> The `contract_call` predicate scope is designed to allow you to use Chainhook to observe direct functions calls within a smart contract. The payload returned by Chainhook will contain transaction data describing the on-chain events that this function elicits.

Reviewer: ```suggestion
The `contract_call` predicate scope is designed to target direct function calls within a smart contract. When triggered, Chainhook will return a payload with transaction data detailing the on-chain events contained in these functions.
```

> By designing your smart contract functions in conjunction with your data source and use case, Chainhook and `contract_call` can become an excellent tool for monitoring interactions with critical contract functionality.

Reviewer: this paragraph is a bit confusing, perhaps its unneeded or needs to be re-worded?

Author: Resolved in latest commit.

<Callout type="warn" title="Requirements">
  To use Chainhook you can either install it
  [directly](/stacks/chainhook/installation) or use [Hiro
  Platform](https://platform.hiro.so/). If you are installing Chainhook, this
  [quickstart](/stacks/chainhook/quickstart) will walk you through setting up a
  predicate file and using it to scan for events.
</Callout>

In this guide, you will learn how to:

1. [Construct a Stacks blockchain Chainhook predicate for a contract call scope](#creating-the-predicate).
2. [Run your Chainhook predicate to generate a payload](#running-the-predicate).
3. [Use a Clarity function to return specific contract data](#return-contract-data-with-the-clarity-function).
4. [Find data related to the contract call within the Chainhook payload](#chainhook-payload).

<Steps>
<Step>

## Creating the predicate

> The `predicate` json object is the primary tool for interacting with the Chainhook data indexer. Chainhook will evaluate our `predicate` to determine what blockchain, network and scope it will use to observe matching transactions.

Reviewer: ```suggestion
The `predicate` is your main interface for querying the Chainhook data indexer. Chainhook uses your specified `predicate` to select the appropriate blockchain, network, and scope for monitoring transactions.
```

Run the below command to generate a template to help you generate your Stacks predicate.

```console
chainhook predicates new contract-call-chainhook.json --stacks
```

<Callout type="info" title="Note">
  Alternatively, [Hiro Platform](https://platform.hiro.so/) has an excellent UI
  to help you to create a `predicate` using a form builder or upload a json file
  containing your `predicate`.
</Callout>

> You can view the full predicate example will be using in this guide below.

Reviewer: ```suggestion
You can view the full predicate example we will be using in this guide below.
```

```json title="contract-call-chainhook.json"
// [!code word:if_this]
{
  "if_this": {
    "scope": "contract_call",
    "contract_identifier": "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV.simple-vote-v0",
    "method": "cast-vote"
  }
}
```

<Callout type="warn" title="Warning">
  The function defined in the `method` must be directly called for Chainhook to
  observe events. Calling the function from another contract or from within a
  different function on the same contract will not generate a `payload`.

```clarity
(define-public (call-cast-vote)
  (cast-vote)
)
```

</Callout>

<Callout type="info">
  You can view other `scope` definitions on the
  [Scopes](/stacks/chainhook/references/scopes) reference page.
</Callout>

### Defining the payload destination with the then_that specification

> Finally, we will define how Chainhook should deliver its data payload when it is triggered by our `predicate` within the `then_that` specification.

Reviewer: ```suggestion
Finally, you need to define how Chainhook delivers the payload when it is triggered by our `predicate` within the `then_that` specification.
```

> When using `file_append`, we specify the `path` we want Chainhook to write its data payload to.

Reviewer: ```suggestion
When using `file_append`, specify the `path` you want Chainhook to post the payload data to.
```

```json title="contract-call-chainhook.json"
// [!code word:then_that]
{
  "then_that": {
    "file_append": {
      "path": "/tmp/events.json"
    }
  }
}
```

<Callout type="info">
  Chainhook is unable to write to a file when using `devnet`.
</Callout>

Chainhook can also post its payload to an endpoint using `http_post`, allowing us to specify the endpoint's `url` and `authorization_header`.

```json title="contract-call-chainhook.json"
// [!code word:then_that]
"then_that": {
        "http_post": {
          "url": "https://webhook.site/abc123456-789e-0fgh-1ijk-23lmno456789",
          "authorization_header": "12345"
        }
      }
```

<Callout type="info" title="Note">
  Chainhook requires `https` to post to your endpoint. You can use a service
  like [LocalTunnel](https://github.com/localtunnel/localtunnel) to test locally
  or a site like [WebhookSite](https://webhook.site).
</Callout>

</Step>
<Step>

## Running the predicate

> Now that we have built our Stacks blockchain `predicate`, we are ready to use Chainhook to scan for blocks that match our `contract_call` scope and view the payload it returns. Whenever this method is directly called within a transaction, Chainhook will observe this event and generate a block level payload with transaction data.

Reviewer: ```suggestion
With our `predicate` set up, we can now use Chainhook to monitor blocks that match our `contract_call` scope and analyze the returned payload. Chainhook will track events where this function is directly invoked in a transaction and generate a detailed payload containing transaction data at the block level.
```

> If you previously installed Chainhook, we will specify our `contract-call-chainhook.json` predicate file in the following command to scan the Stacks blockchain.

Reviewer: > If you previously installed Chainhook

this feels a bit redundant, given that installation has been called out as a requirement. i would assume the reader is following the guide and carry on with the instructions, which i think helps maintain focus and continuity.

Author: Resolved in latest commit.

```

<Callout type="warn" title="Warning">
  The timestamp returned in this object is not the finalized block time you
  would see in [Stacks Explorer](https://explorer.hiro.so/?chain=testnet), but
  `block_time` returned from the Stacks API [get
  block](/stacks/api/blocks/get-block) endpoint.
</Callout>

### Transaction object

Because Chainhook is triggered on the block level, we will receive a single response that contains data specific to each `transaction` that matches our predicate's `contract_call` scope. To find this data, we start with the `apply` array element of the `payload` object. The single object that makes up the `apply` array contains a child element, the `transactions` array. Every `transaction` will be represents by a single object within this array. This `transaction` object contains its own children elements which can be seen in the example below.

```json
{
  "apply": [
    {
      ...
      "transactions": [
      //transaction object(s)
        {
          "transaction_identifier": { ... },
          "metdata": { ... },
          "operations": [],
        }
      ],
      ...
    }
  ],
  "rollback": [ ... ],
  "chainhook": { ... }
}
```

> With our `transaction` object returned, let's start looking at some of the important data elements we can derive from it. The first element `transcation_identifier` is straightforward, containing a hash value that identifies our transaction.

Reviewer: ```suggestion
Once the `transaction` object is returned, we can begin examining important data elements it contains. The first element, `transaction_identifier`, includes a hash value that uniquely identifies our transaction.
```

```json title="contract-call-payload.json"
"transaction_identifier": {
  "hash": "0x98195af8f888d2f9ca3462c41c1691e7798ea6d9e5e3afe42955c0921f981f2c"
}
```

<Callout type="info" title="Note">
  The `operations` array in our example is empty, as our function does not
  involve the transfer of any assets. When assets are transferred the
  `operations` array would contain a record of every `debit` and `credit` of
  those assets as functions are executed by smart contracts. These records are
  not exclusive to a specific predicate scope.
</Callout>

> The next element we will look for data related to our `contract_call` is the `metadata` object and its children elements. Since we are looking to build out our data source, knowing the success state of our transaction will be important. Transactions are broadcasted and observed by Chainhook regardless of whether or not it is successfull, so we will use the the `success` object and its boolean value. We can also derive the transaction `sender` and the contract's returned `result`.

Reviewer: ```suggestion
Next, focus on the `metadata` object within our `contract_call` data. It's crucial to determine the success state of our transaction. Chainhook captures and reports on transactions regardless of their outcome. Utilize the `success` object to assess transaction success. Additionally, extract the sender of the transaction and the `result` returned by the contract.
```

> The next element we will look for data related to our `contract_call` is the `metadata` object and its children elements. Since we are looking to build out our data source, knowing the success state of our transaction will be important. Transactions are broadcasted and observed by Chainhook regardless of whether or not it is successfull, so we will use the the `success` object and its boolean value. We can also derive the transaction `sender` and the contract's returned `result`.

Reviewer: ```suggestion
Next, focus on the `metadata` object within your `contract_call` data. It's crucial to determine the success state of your transaction. Chainhook captures and reports on transactions regardless of their outcome. Utilize the `success` object to assess transaction success. Additionally, extract the sender of the transaction and the `result` returned by the contract.
```

```json title="contract-call-payload.json"
"metadata": {
  ...
  "success": true,
  "sender": "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV",
  "result": "(ok \"Vote cast successfully\")"
  ...
}
```

> Next, we will look at our `kind` object and its elements. Within the `data` object, we find data related to the contracts our function exists on as a value of `contract_identifier`, the function called on that contract as a value of `method` and the arguments passed to this function when invoking it as an array of `args`. In this instance, our `cast-vote` function does not accept any arguements so the `args` array is empty.

Reviewer: ```suggestion
Moving forward, examine the `kind` object and its components within your `contract_call`. The data object contains crucial information: `contract_identifier` specifies the contract our function resides on, `method` identifies the function called, and `args` lists the arguments passed when invoking this function. In our case, the `cast-vote` function accepts no arguments, resulting in an empty `args` array.
```

```json title="contract-call-payload.json"
// [!code word:kind]
{
  "metadata":
  ...
  "kind": {
    "data": {
      "args": [],
      "contract_identifier": "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV.simple-vote-v0",
      "method": "cast-vote"
    }
    ...
  }
  ...
}
```

> The last element of the `metadata` object we will examine is the `receipt` object and its `events` array and its object elements. In our `contract_call` example, the information we want to pull into our data source will be found in the `data` object of these elements. Like the `kind` object above, our `data` object also contains a `contract_identifer` key with our `simple-vote-v0` contract as its value. Because we are returning data from our contract's `cast-vote` function using `print`, we will find the `topic` and `value` keys. `Topic` shows us that we are using `print` to return the data set in our function and the `value` returns the data we defined in that `print`.

Reviewer: ```suggestion
The final element to analyze in the `metadata` object is the `receipt` object, particularly its `events` array and the objects within it.

For your `contract_call`, the relevant data is housed in the `data` object of these elements. Similar to the `kind` object, our `data` object includes a `contract_identifier` key, which in this case is linked to the `simple-vote-v0` contract.

Since you are using the `print` statement to return data from your contract's `cast-vote` function, the `events` array will contain `topic` and `value` keys.

The `topic` indicates that the print statement is being used to output data, while the `value` key will display the specific data set within that `print` function.
```

```json title="contract-call-payload.json"
// [!code word:receipt]
{
  "metadata":
  ...
  "receipt": {
    "contract_calls_stack": [],
    "events": [
      {
        "data": {
          "contract_identifier": "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV.simple-vote-v0",
          "topic": "print",
          "value": {
            "notification": "cast-vote",
            "payload": {
              "status": "Has voted set to true",
              "totalVotes": 1,
              "voter": "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV"
            }
          }
        },
        "position": {
          "index": 0
        },
        "type": "SmartContractEvent"
      }
    ],
    "mutated_assets_radius": [],
    "mutated_contracts_radius": [
      "STJ81C2WPQHFB6XTG518JKPABWM639R2X37VFKJV.simple-vote-v0"
    ]
  }
  ...
}
```

</Step>
</Steps>

---

> Now that we have identified where the relevant data lives in the Chainhook payload, you can build the necessary logic into your API to extract it. You should be able to build a data source for your DApp based on the `contract_call` predicate that looks like the example below.

Reviewer: ```suggestion
Now that you've located the relevant data in the payload, you can start to extract the relevant information into your API.

The following is an example of how you might store your information in a database table:
```
